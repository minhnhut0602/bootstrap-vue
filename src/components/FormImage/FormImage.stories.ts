import FormImage from '.';
import {ref} from 'vue';

export default {
    title: 'Forms/Form image',
    component: FormImage,
    argTypes: {},
};

const template = (args) => ({
    components: {FormImage},
    setup: () => ({args, value: ref<File>()}),
    template: '<FormImage v-model="value" v-bind="args" />',
});

export const Default = template.bind({});
