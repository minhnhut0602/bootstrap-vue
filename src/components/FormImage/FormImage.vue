<template>
    <label
        class="form-image"
        :for="id"
        :style="style"
        @drop="onDrop"
        @dragenter="cancelDrag"
        @dragover="cancelDrag"
    >
        <input
            :id="id"
            ref="input"
            accept="image/*"
            hidden
            type="file"
            @change="onChange"
        >

        <div
            v-if="preview"
            class="form-image-btn-edit"
        />
    </label>
</template>

<script lang="ts">
import {PropType, computed, ref, watch} from 'vue';
import {idProps} from '@/composables/useId';

export const dataTypes = ['file', 'base64'] as const;

export type DataType = typeof dataTypes[number];
</script>

<script lang="ts" setup>
const props = defineProps({
    dataType: {
        type: String as PropType<DataType>,
        default: 'base64',
    },
    ...idProps,
    modelValue: {
        type: [Object, String] as PropType<File|string>,
        default: undefined,
    },
    preview: {
        type: String,
        default: undefined,
    },
});

const emit = defineEmits<{
  (e: 'update:modelValue', v: string|File): void,
}>();

const input = ref<HTMLInputElement>();

const preview = ref<string|undefined>(props.preview);

const file = ref<File|undefined>();

const style = computed(() => {
    if (!preview.value) {
        return {};
    }

    return {
        'background-color': 'transparent',
        'background-image': `url(${preview.value})`,
        'background-size': 'cover',
    };
});

const value = computed<string|File|undefined>({
    get: () => props.modelValue,
    set: (v?: string|File) => emit('update:modelValue', v),
});

watch(file, f => readFile(f));

const cancelDrag = (e: Event) => e.preventDefault();

const onChange = (e: {target: HTMLInputElement}) => {
    const newFile = e.target.files.item(0);

    if (newFile?.type.match('image.*')) {
        file.value = newFile;

        return true;
    }

    return false;
};

const onDrop = (e: DragEvent) => {
    e.preventDefault();

    input.value.files = e.dataTransfer.files;

    onChange({target: input.value});
};

const readFile = (file: File) => {
    if (!file) {
        return value.value = undefined;
    }

    if (props.dataType === 'file') {
        value.value = file;
    }

    const reader = new FileReader();

    reader.onload = () => {
        preview.value = reader.result as string;

        if (props.dataType === 'base64') {
            value.value = reader.result as string;
        }
    };

    reader.readAsDataURL(file);

    return reader;
};
</script>
