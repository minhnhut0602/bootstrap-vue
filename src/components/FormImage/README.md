# FormImage

```vue live
<FormImage>Default Example Usage</FormImage>
```

## Props

| Prop name  | Description | Type         | Values | Default   |
| ---------- | ----------- | ------------ | ------ | --------- |
| dataType   |             | DataType     | -      | 'base64'  |
| modelValue |             | File\|string | -      | undefined |
| preview    |             | string       | -      | undefined |

## Slots

| Name        | Description | Bindings |
| ----------- | ----------- | -------- |
| placeholder |             |          |

## Events

| Event name        | Properties | Description |
| ----------------- | ---------- | ----------- |
| update:modelValue |            |
