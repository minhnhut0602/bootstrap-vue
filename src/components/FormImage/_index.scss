.form-image {
    position: relative;
    display: block;
    @include border-radius($input-border-radius);
    border: $input-border-width solid $input-border-color;
    width: 4rem;
    height: 4rem;
    cursor: pointer;
    background-color: $input-bg;
    background-image: escape-svg(url("data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' fill='#{$input-border-color}' viewBox='0 0 16 16' width='16' height='16'><path d='M.5 9.9a.5.5 0 0 1 .5.5v2.5a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1v-2.5a.5.5 0 0 1 1 0v2.5a2 2 0 0 1-2 2H2a2 2 0 0 1-2-2v-2.5a.5.5 0 0 1 .5-.5z'/><path d='M7.646 1.146a.5.5 0 0 1 .708 0l3 3a.5.5 0 0 1-.708.708L8.5 2.707V11.5a.5.5 0 0 1-1 0V2.707L5.354 4.854a.5.5 0 1 1-.708-.708l3-3z'/></svg>"));
    background-position: center;
    background-repeat: no-repeat;
    background-size: 1rem;

    .form-image-btn-edit {
        position: absolute;
        top: $input-padding-y;
        right: $input-padding-y;
    }
}

.form-image-btn-edit {
    @include border-radius($btn-border-radius-sm);
    width: 1rem;
    height: 1rem;
    background-color: rgba($dark, .8);
    background-image: escape-svg(url("data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' width='16' height='16' fill='#{$white}' viewBox='0 0 16 16'><path d='M12.854.146a.5.5 0 0 0-.707 0L10.5 1.793 14.207 5.5l1.647-1.646a.5.5 0 0 0 0-.708l-3-3zm.646 6.061L9.793 2.5 3.293 9H3.5a.5.5 0 0 1 .5.5v.5h.5a.5.5 0 0 1 .5.5v.5h.5a.5.5 0 0 1 .5.5v.5h.5a.5.5 0 0 1 .5.5v.207l6.5-6.5zm-7.468 7.468A.5.5 0 0 1 6 13.5V13h-.5a.5.5 0 0 1-.5-.5V12h-.5a.5.5 0 0 1-.5-.5V11h-.5a.5.5 0 0 1-.5-.5V10h-.5a.499.499 0 0 1-.175-.032l-.179.178a.5.5 0 0 0-.11.168l-2 5a.5.5 0 0 0 .65.65l5-2a.5.5 0 0 0 .168-.11l.178-.178z'/></svg>"));
    background-position: center;
    background-repeat: no-repeat;
    background-size: .5rem;
}
