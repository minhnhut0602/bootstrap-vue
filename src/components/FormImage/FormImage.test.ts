import FormImage from '.';
import {shallowMount} from '@vue/test-utils';
import { nextTick } from '@vue/runtime-core';

describe('template', () => {
    componentRenderTest(FormImage, {
        // Force an id to prevent snapshots from failing because of uuid.
        props: {id: 'dont-change-please'},
    });

    it('renders btn edit when has preview', () => {
        const wrapper = shallowMount(FormImage, {props: {preview: 'some-preview'}});

        expect(wrapper.find('.form-image-btn-edit').exists()).toBe(true);
    });

    // TODO: Unable to test drop, dragenter, dragover properly. It is currently
    // not possible to create a FileList instance used during the assigment of
    // the files to a file input element. It is currently not possible to mock
    // methods defined in setup.
});

describe('file', () => {
    describe('watch', () => {
        it('sets value to undefined when not thruthful', async () => {
            const wrapper = shallowMount(FormImage, {props: {modelValue: 'some-file'}});

            // First change to something that is not undefined but still falsly
            // to prevent the watcher to read the value as a file.
            wrapper.vm.file = false;

            await nextTick();

            wrapper.vm.file = undefined;

            // Cannot test value to be exactly undefined since it is a computed
            // property that will emit when set.
            expect(wrapper.emitted('update:modelValue')[0]).toEqual([undefined]);
        });

        // More detailed tests are done in readFile tests.
    });
});

describe('style', () => {
    it('is empty when there is no preview', () => {
        const wrapper = shallowMount(FormImage);

        expect(wrapper.vm.style).toEqual({});
    });

    it('has background styling when there is a preview', () => {
        const wrapper = shallowMount(FormImage);

        wrapper.vm.preview = 'some-preview';

        expect(wrapper.vm.style).toEqual({
            'background-color': 'transparent',
            'background-image': 'url(some-preview)',
            'background-size': 'cover',
        });
    });
});

describe('value', () => {
    it('is equal to modelValue', () => {
        const wrapper = shallowMount(
            FormImage,
            {props: {modelValue: 'some-file'}},
        );

        expect(wrapper.vm.value).toBe('some-file');
    });

    it('emits update:modelValue when set', () => {
        const wrapper = shallowMount(FormImage);

        wrapper.vm.value = 'some-value';

        expect(wrapper.emitted('update:modelValue')[0]).toEqual(['some-value']);
    });
});

describe('readFile', () => {
    // The truthful test is done in watcher test of file to test if this method
    // is called.

    it('handles when dataType is file', async () => {
        const readAsDataURLSpy = jest.spyOn(FileReader.prototype, 'readAsDataURL');

        const wrapper = shallowMount(FormImage, {props: {dataType: 'file'}});

        const file = new File([], 'test.png');

        const reader = wrapper.vm.readFile(file);

        // Cannot test value to be exactly undefined since it is a computed
        // property that will emit when set.
        expect(wrapper.emitted('update:modelValue')[0]).toEqual([file]);

        expect(readAsDataURLSpy).toBeCalledWith(file);

        Object.defineProperty(reader, 'result', {value: 'some-preview'});

        reader.onload();

        expect(wrapper.vm.preview).toBe('some-preview');
    });

    it('handles when dataType is base64', async () => {
        const readAsDataURLSpy = jest.spyOn(FileReader.prototype, 'readAsDataURL');

        const wrapper = shallowMount(FormImage, {props: {dataType: 'base64'}});

        const file = new File([], 'test.png');

        const reader = wrapper.vm.readFile(file);

        expect(readAsDataURLSpy).toBeCalledWith(file);

        Object.defineProperty(reader, 'result', {value: 'some-preview'});

        reader.onload();

        expect(wrapper.vm.preview).toBe('some-preview');

        // Cannot test value to be exactly undefined since it is a computed
        // property that will emit when set.
        expect(wrapper.emitted('update:modelValue')[0]).toEqual(['some-preview']);
    });
});
