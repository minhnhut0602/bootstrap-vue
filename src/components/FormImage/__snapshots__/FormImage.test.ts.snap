// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template renders default 1`] = `"<label class=\\"form-image\\" for=\\"dont-change-please\\"><input id=\\"dont-change-please\\" accept=\\"image/*\\" hidden=\\"\\" type=\\"file\\"></label>"`;
 